### PR TITLE
[ResourceIsolation] Bind Workgroup to PipelineDriver

### LINKED ISSUE #2689: Bind pipeline drivers to workgroups
## Enhancement

Each query belongs to a certain workgroups, so bind pipeline drivers originates from the same query to the workgroup. when prepare FragmentExecutor for a fragment instance, the MemTracker of the target workgroup is used to initialize the MemTracker of RuntimeState.  in addition, 
1. drivers have a field _workgroup to keep its target workgroup. 
2. Initialization of workgroups for verification moved to ExecEnv::init since MemTracker of workgroups are children of query_pool_mem_tracker of ExecEnv and the latter should be initialized first.

### PR DESCRIPTION
**Pull request by satanson**
## Enhancement

Each query belongs to a certain workgroups, so bind pipeline drivers originates from the same query to the workgroup. when prepare FragmentExecutor for a fragment instance, the MemTracker of the target workgroup is used to initialize the MemTracker of RuntimeState. in addition,

1. drivers have a field _workgroup to keep its target workgroup.
2. Initialization of workgroups for verification moved to ExecEnv::init since MemTracker of workgroups are children of query_pool_mem_tracker of ExecEnv and the latter should be initialized first.

### CODE REVIEW

File: be/src/exec/workgroup/work_group.cpp [hunk 1]
```diff
@@ -59,16 +101,17 @@ WorkGroupQueue& WorkGroupManager::get_io_queue() {
...
-public:
-    DefaultWorkGroupInitialization() {
-        auto default_wg = std::make_shared<WorkGroup>("default_wg", 1, 10, 10, 10, WorkGroupType::WG_DEFAULT);
-        default_wg->init();
-        WorkGroupManager::instance()->add_workgroup(default_wg);
-    }
-} _;
-} // namespace
+DefaultWorkGroupInitialization::DefaultWorkGroupInitialization() {
+    auto default_wg = std::make_shared<WorkGroup>("default_wg", 0, 1, 20L * (1L << 32), 10, WorkGroupType::WG_DEFAULT);
```
Comment (ZiheLiu) [inline be/src/exec/workgroup/work_group.cpp:105]: `1L << 32` is 4GB, maybe it should be `1L << 30`?

Comment (satanson) [reply be/src/exec/workgroup/work_group.cpp:105]: ok

File: gensrc/thrift/InternalService.thrift [hunk 2]
```diff
@@ -249,6 +249,21 @@ enum InternalServiceVersion {
...
 }
 
+enum TWorkGroupType {
+    WG_NORMAL,
+    WG_DEFAULT,
+    WG_REALTIME
+}
+
+struct TWorkGroup {
+    1: required i32 id
```
Comment (kangkaisen) [inline gensrc/thrift/InternalService.thrift:259]: always use optional, shouldn't use required.

File: be/src/exec/pipeline/fragment_executor.cpp [hunk 3]
```diff
@@ -104,12 +109,20 @@ Status FragmentExecutor::prepare(ExecEnv* exec_env, const TExecPlanFragmentParam
 
     LOG(INFO) << "Prepare(): query_id=" << print_id(query_id)
               << " fragment_instance_id=" << print_id(params.fragment_instance_id) << " backend_num=" << backend_num;
+    WorkGroupPtr wg;
+    if (request.__isset.workgroup && request.workgroup.id != WorkGroup::DEFAULT_WG_ID) {
+        wg = std::make_shared<WorkGroup>(request.workgroup);
+        wg = WorkGroupManager::instance()->add_workgroup(wg);
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/fragment_executor.cpp:115]: If user sets the session variable `workgroup_id` as an unknown number, we will create a workgroup in the BE here.
Should we use the default workgroup in the case, or FE will guarantee that the `workgroup_id` must exist?

File: gensrc/thrift/InternalService.thrift (hunk 2 above)
Comment (satanson) [reply gensrc/thrift/InternalService.thrift:259]: ok

File: be/src/exec/pipeline/fragment_executor.cpp (hunk 3 above)
Comment (satanson) [reply be/src/exec/pipeline/fragment_executor.cpp:115]: ok

Comment (satanson) [reply be/src/exec/pipeline/fragment_executor.cpp:115]: workgoup is pigbacked by TExecPlanFragmentParams, it maybe absent. so for the first time that workgroup is delivered to BE, BE create workgroup and register it.

**Review by ZiheLiu — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by laotan332 — APPROVED** [review]

**Pull request merged**